rfc/merkle-structure.md

```diff
@@ -0,0 +1,320 @@
+# Abstract
+
+Inter Planetary Linked Data (IPLD) is a fundamental building block of the IPFS. IPLD Links (a.k.a CIDs) are quintessential, yet I'd argue a major design flaw (on par with null pointers ?)
```

Comment:
> Links (a.k.a CIDs) are quintessential, yet I'd argue a major design flaw (on par with null pointers ?)

Woah, I could not disagree more. Yes, you have to reason about the network, _but this is explicitly a distributed system_. We can't reason about the tradeoffs the same as a local structure, and trying to merge the two always has tradeoffs.

> Much of the pain in traditional distributed programming comes from this mismatch:  programmers are expected to bridge from an ordered programming model into a disordered reality that executes their code.
> 
> — BloomLang website

The four approaches are:

1. Try to make everything feel local even if remote
  * Fails because not everything is actually local
  * Simple in simple cases, extremely confusing and frustrating in complex ones
2. Try to make everything feel remote even if local
  * More-or-less "like Erlang"
  * Simple cases always have to handle the "what if it's not there" case, but complex cases are consistent
3. Separate the cases
  * Have a distinction between things that are expected to be together versus could be unlinked (i.e. like most of the internet / HTTP _and also_ IPLD)
4. Monotonic / eventual consistency
  * i.e. Who cares if the remote doc isn't available, you can work with partial information just fine and progressively enhance 

I would generally claim that `#1` is not a fit for distributed systems — full stop.

`#2` is very workable, but inconvenient (programmers are unhappy about the unnecessary complexity when they _know_ something is local).

`#3` is what most systems do today, It's an extremely successful model. With Merklized data, I think the problem comes up when _humans_ don't want to work with pointer indirection or you have a hard requirement to access data that you can't fetch / data availability. I would argue that data availability should be solved at a different layer (like they do on the the non-Merklized web — AWS has many products aimed at this problem). Human factors can be solved many ways (such as inline-ipld, but the class of solution is certainly broader than that).

`#4` is kind of subtype / solution to `#3`. If you're in an eventually consistent domain (which sure, many things are not, e.g. streaming a movie), then you can largely side step the issue.

Reply:
Without going too deep into a pure theory discussion, I'd also like to gesture at Mark Miller's Vat concept as another example of `#3`, and maybe in which some ideas could be pulled.

From his thesis "Robust Composition", which is more about a programming model than a storage layer (but you can adapt the ideas):

There's obviously a distinction between local and remote:

<img width="697" alt="Screenshot 2024-02-03 at 12 13 50" src="https://github.com/web3-storage/RFC/assets/1052016/0248d95d-224b-4051-895b-67a8cbafa8d9">

Local calls stay inside a Vat (in his terminology), which is a sequentially executing VM/container that guarantees that ant agent inside it can call an of the others inside it, in a totally predictable order. Calling across Vats may fail or introduce nondeterministic asynchrony. Two Vats can be on the same machine, or networked — they're just an abstraction of how you think some agents will expect others to be guaranteed responsive not.

<img width="631" alt="Screenshot 2024-02-03 at 12 14 21" src="https://github.com/web3-storage/RFC/assets/1052016/298ccdab-bfe8-4a59-8e5e-8239c45272d6">.

I think of IPLD this way today ("these fields all are guaranteed to move together"), but I wonder if there's other ways to slice the problem that we could borrow from the Vat idea?

Reply:
I'm not disagreeing with what you are saying here. That said, what I was trying to convey is that backing whether something is local or not into data identifier is not a good idea.

What's described in the proposal is probably `#4` or perhaps `#2`, definitely not `#1`. Hopefully newer version conveys the point more clearly.